File: backend/MediaHub.Api/Models/Dtos.cs
namespace MediaHub.Api.Models;

// Response/request shapes mirror the contract documented in the repo root README.
// System.Text.Json serializes these as camelCase (configured in Program.cs).

// SizeBytes is appended last (additive): new optional fields go at the end so the
// record's existing positional shape is untouched. JSON is matched by name regardless.
public sealed record VideoSummaryDto(
    string Id,
    string Title,
    string? Description,
    string? ThumbnailUrl,
    int? DurationSeconds,
    DateTimeOffset CreatedAt,
    long? SizeBytes = null);

public sealed record VideoListDto(IReadOnlyList<VideoSummaryDto> Videos);

public sealed record VideoDetailDto(
    string Id,
    string Title,
    string? Description,
    string? ThumbnailUrl,
    int? DurationSeconds,
    string PlaybackUrl,
    DateTimeOffset PlaybackUrlExpiresAt,
    string MimeType,
    DateTimeOffset CreatedAt,
    long? SizeBytes = null);

/// <summary>JSON body for registering a video that already exists in R2.</summary>
public sealed record CreateVideoRequest(
    string Title,
    string? Description,
    string ObjectKey,
    string? ThumbnailUrl,
    int? DurationSeconds,
    string? MimeType,
    long? SizeBytes = null);

public sealed record AppReleaseDto(
    int VersionCode,
    string VersionName,
    string Notes,
    string DownloadUrl,
    long SizeBytes,
    string Sha256,
    int MinSdk,
    DateTimeOffset PublishedAt,
    bool ForceUpdate);

/// <summary>
/// Whether the app must present an access code, and whether the one it sent (header
/// <c>X-Access-Code</c>) is currently valid. <c>required=false</c> means the gate is off and the
/// app should proceed straight to content.
/// </summary>
public sealed record AccessStatusDto(bool Required, bool Valid);

File: android-tv/app/src/main/java/com/tvvideohub/tv/data/dto/Dtos.kt
package com.tvvideohub.tv.data.dto

import kotlinx.serialization.Serializable

/**
 * DTOs mirroring the backend HTTP/JSON contract (see repo root README.md, "API contract v1").
 * Field names MUST match the JSON exactly. Nullable fields are nullable here too.
 */

/** Wrapper for `GET /api/videos`. */
@Serializable
data class VideoListResponse(
    val videos: List<VideoSummary> = emptyList()
)

/** One item in `GET /api/videos`. */
@Serializable
data class VideoSummary(
    val id: String,
    val title: String,
    val description: String = "",
    val thumbnailUrl: String? = null,
    val durationSeconds: Int? = null,
    val createdAt: String
)

/** `GET /api/videos/{id}` — details plus a short-lived playback URL. */
@Serializable
data class VideoDetail(
    val id: String,
    val title: String,
    val description: String = "",
    val thumbnailUrl: String? = null,
    val durationSeconds: Int? = null,
    val playbackUrl: String,
    val playbackUrlExpiresAt: String,
    val mimeType: String,
    val createdAt: String
)

/**
 * `GET /api/app/latest` — newest published APK.
 * Note: the endpoint returns HTTP 204 (no body) when no release exists yet;
 * the API layer maps that to a null AppRelease.
 */
@Serializable
data class AppRelease(
    val versionCode: Int,
    val versionName: String,
    val notes: String = "",
    val downloadUrl: String,
    val sizeBytes: Long,
    val sha256: String,
    val minSdk: Int,
    val publishedAt: String,
    /** When true the update is mandatory: the app must not let the user dismiss/cancel it. */
    val forceUpdate: Boolean = false
)

/**
 * `GET /api/app/access` — whether the app must present an access code, and whether the code it
 * sent (the `X-Access-Code` header) is currently valid. `required=false` ⇒ gate off, proceed.
 */
@Serializable
data class AccessStatus(
    val required: Boolean = false,
    val valid: Boolean = false
)
